core/client/fs/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -172,21 +210,38 @@ private int positionedReadInternal(long pos, byte[] b, int off, int len) throws
     }
 
     int lenCopy = len;
+    int maxRetry = Configuration.getInt(PropertyKey.USER_BLOCK_WORKER_CLIENT_READ_RETRY);
+    int retryLeft = maxRetry;
+    boolean retryCurrentWorker = true;
+    Set<WorkerNetAddress> lostWorkers = new HashSet<>();
     while (len > 0) {
       if (pos >= mLength) {
         break;
       }
       long blockId = mStatus.getBlockIds().get(Math.toIntExact(pos / mBlockSize));
       BlockInStream stream = null;
       try {
-        stream = mBlockStore.getInStream(blockId, mOptions);
+        stream = mBlockStore.getInStream(blockId, mOptions, lostWorkers);
         long offset = pos % mBlockSize;
-        int bytesRead =
-            stream.positionedRead(offset, b, off, (int) Math.min(mBlockSize - offset, len));
-        Preconditions.checkState(bytesRead > 0, "No data is read before EOF");
-        pos += bytesRead;
-        off += bytesRead;
-        len -= bytesRead;
+        try {
+          int bytesRead =
+              stream.positionedRead(offset, b, off, (int) Math.min(mBlockSize - offset, len));
+          Preconditions.checkState(bytesRead > 0, "No data is read before EOF");
+          pos += bytesRead;
+          off += bytesRead;
+          len -= bytesRead;
+          retryLeft = maxRetry;
+          retryCurrentWorker = true;
+        } catch (AbortedException | DeadlineExceededException | ConnectException e) {
+          if (!retryCurrentWorker) {
+            if (retryLeft <= 0) {
+              throw e;
+            }
+            retryLeft--;
+          }
+          stream = prepareForRetry(stream, e, lostWorkers, retryCurrentWorker);
```

Comment:
on exception case, do we simply reconstruct the stream?
don't we need to also try line 227-232?

Reply:
When one of the above exceptions is thrown, this function either
- reconstruct the stream from same parameters on first time to retry from the same worker
- or simply set the stream to null and put current worker to `lostWorkers` list and wait for the next loop iteration to choose a new worker and build a new stream.

Either way `227-232` is retried on the next iteration.